src/runtime/hooks.ts

```diff
@@ -36,8 +36,15 @@ export function useEnv<E extends Env>(): E {
 
 function extendEnv(currentEnv: Object, extension: Object): Object {
   const env = Object.create(currentEnv);
-  const descrs = Object.getOwnPropertyDescriptors(extension);
-  return Object.freeze(Object.defineProperties(env, descrs));
+  while (extension !== Object.prototype) {
+    const descrs = Object.getOwnPropertyDescriptors(extension);
+    Object.defineProperties(env, descrs);
```

Comment:
Oh and because we're doing this while climbing the prototype chain, properties on the prototype will shadow properties on the object instead of the other way around which is undesirable in any case.